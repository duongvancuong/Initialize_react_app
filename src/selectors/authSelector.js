export const getAuthentication = state => (state.auth.isAuthenticated);
export const getToken = state => (state.auth.token);
