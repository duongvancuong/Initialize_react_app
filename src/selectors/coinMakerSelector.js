/* eslint-disable */
export const getCryptocurrencyLatest = state => {
  const datas = state.cryptocurrency.data;
  const columns = ['name', 'market_cap', 'price', 'volume_24h', 'circulating_supply', 'percent_change_24h'];
  const rows = datas.map(data => {
    return {
      'name': data.name,
      'market_cap': data.quote.USD.market_cap,
      'price': data.quote.USD.price,
      'volume_24h': data.quote.USD.volume_24h,
      'circulating_supply': data.circulating_supply,
      'percent_change_24h': data.quote.USD.percent_change_24h,
    }
  });
  return {
    columns: columns,
    rows: rows
  }
}
