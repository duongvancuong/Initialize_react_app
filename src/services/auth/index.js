import { UnauthenticatedRequest, AutheticatedRequest } from '../api';
import config from '../../config';

const URL_LOGIN = '/auth/login';
const URL_REGISTER = '/signup';
const URL_LOGOUT = '/auth/logout';

const CONFIG_OPTION = {
  baseURL: config.baseURL,
  timeout: 10000,
  params: {},
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
};

export const login = data => UnauthenticatedRequest(CONFIG_OPTION).post({
  url: URL_LOGIN,
  data,
}).then(res => res.data).catch(error => error.response.data);

export const register = data => UnauthenticatedRequest(CONFIG_OPTION).post({
  url: URL_REGISTER,
  data,
}).then(res => res.data).catch(error => error.response.data);

export const logout = (token) => {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    Authorization: token,
  };
  const CONFIG_OPTION_AUTHENTICATION = { ...CONFIG_OPTION, headers };
  return AutheticatedRequest(CONFIG_OPTION_AUTHENTICATION).delete({
    url: URL_LOGOUT,
  }).then(res => res.data).catch(error => error.response.data);
};
