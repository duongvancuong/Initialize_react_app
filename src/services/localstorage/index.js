import _ from 'lodash';

const localStorageKey = 'react_app_state';

export const saveLocalStorage = (key, value) => {
  try {
    localStorage.setItem(key, JSON.stringify(value));
  } catch (err) {
    // Nothing to do
  }
};

export const loadLocalStorage = (key) => {
  try {
    return JSON.parse(localStorage.getItem(key));
  } catch (error) {
    return undefined;
  }
};

export const deleteLocalStorage = (key) => {
  localStorage.removeItem(key);
};
