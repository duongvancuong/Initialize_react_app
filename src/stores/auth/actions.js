import { createAction } from 'redux-actions';

export const authenticateUser = createAction('AUTHENTICATE_USER');
export const authenticateUserSuccess = createAction('AUTHENTICATE_USER_SUCCESS');
export const authenticateUserError = createAction('AUTHENTICATE_USER_ERROR');
export const loadingAction = createAction('LOADING_ACTION');
export const logoutUser = createAction('LOGOUT_USER');
export const logoutUserSuccess = createAction('LOGOUT_USER_SUCCESS');
export const cleanErrorLogin = createAction('CLEAN_ERROR_LOGIN');
export const cleanErrorLoginSuccess  = createAction('CLEAN_ERROR_LOGIN_SUCCESS');
