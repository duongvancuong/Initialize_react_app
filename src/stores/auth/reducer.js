import { combineActions, handleActions } from 'redux-actions';

import {
  authenticateUserSuccess,
  authenticateUserError,
  loadingUserData,
  logoutUserSuccess,
  cleanErrorLoginSuccess,
  handleExceptionUserError,
} from './actions';

const initialState = {
  isAuthenticated: false,
  refreshToken: '',
  token: '',
  expiredAt: '',
  isLoading: false,
  error: '',
};

const auth = handleActions({
  [combineActions(
    authenticateUserSuccess,
    authenticateUserError,
    loadingUserData,
    logoutUserSuccess,
    cleanErrorLoginSuccess,
    handleExceptionUserError,
  )]: (state, action) => ({
    ...state, ...action.payload,
  }),
},
initialState);

export default auth;
