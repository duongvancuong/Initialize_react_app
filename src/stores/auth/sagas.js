/* eslint-disable */
import {
  put,
  call,
  takeLatest,
  select,
  all,
} from 'redux-saga/effects';
import { login, logout } from '../../services/auth';

import {
  authenticateUser,
  authenticateUserSuccess,
  authenticateUserError,
  loadingUserData,
  logoutUser,
  logoutUserSuccess,
  cleanErrorLogin,
  cleanErrorLoginSuccess,
  handleExceptionUserError,
} from './actions';

import { getToken } from '../../selectors/authSelector';

const authLogin = token => ({
  isAuthenticated: true,
  token: token.token,
  refreshToken: token.refresh_token,
  expiredAt: token.expired_at,
  error: '',
});

const authLogout = {
  isAuthenticated: false,
  refresh_token: '',
  token: '',
  expired_at: '',
  isLoading: false,
  error: '',
};

export function* loginUserSaga(data) {
  try {
    yield put(loadingUserData({ isLoading: true }));
    const response = yield call(login, data.payload);
    if (response.error_code) {
      yield put(authenticateUserError({ error: `${response.error_code}` }));
    }
    else {
      const payload = authLogin(response.token);
      yield put(authenticateUserSuccess(payload));
    }
    yield put(loadingUserData({ isLoading: false }));
  }
  catch (e) {
    yield put(loadingUserData({ isLoading: false }));
    yield put(handleExceptionUserError({ error: 'Something Wrong!' }));
  }
}

export function* logoutUserSaga() {
  try {
    yield put(loadingUserData({ isLoading: true }));
    const token = yield select(getToken);
    yield call(logout, token);
    yield [
      put(logoutUserSuccess(authLogout)),
    ];
    yield put(loadingUserData({ isLoading: false }));
  }
  catch (error) {
    yield put(loadingUserData({ isLoading: false }));
    yield put(handleExceptionUserError({ error: 'Something Wrong!' }));
  }
}

export function* cleanErrorLoginSaga() {
  try {
    yield put(cleanErrorLoginSuccess({ error: '' }));
  }
  catch (error) {
    yield put(handleExceptionUserError({ error: 'Something Wrong!' }));
  }
}

export default function* root() {
  yield all([
    takeLatest(authenticateUser, loginUserSaga),
    takeLatest(logoutUser, logoutUserSaga),
    takeLatest(cleanErrorLogin, cleanErrorLoginSaga),
  ]);
}
