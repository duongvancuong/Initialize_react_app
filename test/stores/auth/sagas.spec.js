import { expectSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';
import { select } from 'redux-saga/effects';

import root, { loginUserSaga, logoutUserSaga, cleanErrorLoginSaga } from '../../../src/stores/auth/sagas';
import {
  authenticateUserSuccess,
  authenticateUserError,
  loadingUserData,
  logoutUserSuccess,
  cleanErrorLoginSuccess,
  handleExceptionUserError,
} from '../../../src/stores/auth/actions';

import { login, logout } from '../../../src/services/auth';
import { getToken } from '../../../src/selectors/authSelector';

describe('Auth', () => {
  it('should have the expected watchers', done => {
    expectSaga(root)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      })
  });

  it('should fetch authenticate successfully', () => {
    const data = {
      token: {
        token: '123123123123',
        refresh_token: '123123123123123',
        expired_at: '123123123',
      }
    };
    return expectSaga(loginUserSaga, {payload: {email:'test', password: 'test'}})
      .provide([
        [matchers.call.fn(login), data],
      ])
      .dispatch(loadingUserData({ isLoading: true }))
      .put(authenticateUserSuccess({
        isAuthenticated: true,
        token: "123123123123",
        refreshToken: "123123123123123",
        expiredAt: "123123123",
        error: "",
      }))
      .dispatch(loadingUserData({ isLoading: false }))
      .run();
  });

  it('should fetch authenticate failure', () => {
    const errorData = {
      error_code: 602,
      message: 'authentication error',
    };
    return expectSaga(loginUserSaga, {payload: {email:'test', password: 'test'}})
      .provide([
        [matchers.call.fn(login), errorData],
      ])
      .dispatch(loadingUserData({ isLoading: true }))
      .put(authenticateUserError({error: '602'}))
      .dispatch(loadingUserData({ isLoading: false }))
      .run();
  });

  it('should fetch authenticate Error Saga', () => {
    const error = new Error('something went wrong');

    return expectSaga(loginUserSaga)
      .provide([
        [matchers.call.fn(login), throwError(error)],
      ])
      .dispatch(loadingUserData({ isLoading: true }))
      .put(handleExceptionUserError({error: 'something went wrong'}))
      .dispatch(loadingUserData({ isLoading: false }))
      .run();
  });

  it('should logout user', () => {
    return expectSaga(logoutUserSaga)
      .provide([
        [select(getToken), '123123123123'],
        [matchers.call.fn(logout)],
      ])
      .dispatch(loadingUserData({ isLoading: true }))
      .put(logoutUserSuccess({
        isAuthenticated: false,
        refresh_token: '',
        token: '',
        expired_at: '',
        isLoading: false,
        error: '',
      }))
      .dispatch(loadingUserData({ isLoading: false }))
      .run();
  });

  it('should logout user failure', () => {
    const error = new Error('something went wrong');

    return expectSaga(logoutUserSaga)
      .provide([
        [matchers.call.fn(logout), throwError(error)],
      ])
      .dispatch(loadingUserData({ isLoading: true }))
      .put(handleExceptionUserError({error: 'something went wrong'}))
      .dispatch(loadingUserData({ isLoading: false }))
      .run();
  });

  it('should clean error login successfully', () => {
    return expectSaga(cleanErrorLoginSaga)
      .dispatch(loadingUserData({ isLoading: true }))
      .put(cleanErrorLoginSuccess({ error: '' }))
      .dispatch(loadingUserData({ isLoading: false }))
      .run();
  });
});
