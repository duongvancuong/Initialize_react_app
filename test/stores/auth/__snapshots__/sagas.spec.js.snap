// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "@@redux-saga-test-plan/json/function/actionCreator",
          "@@redux-saga-test-plan/json/function/loginUserSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "@@redux-saga-test-plan/json/function/actionCreator",
          "@@redux-saga-test-plan/json/function/logoutUserSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "@@redux-saga-test-plan/json/function/actionCreator",
          "@@redux-saga-test-plan/json/function/cleanErrorLoginSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;
